jobbergateapi2/tests/apps/application_permissions/test_routers.py

```diff
@@ -0,0 +1,247 @@
+async def test_create_application_permissions_duplicated(user_data, client):
+    """
+    This test proves that is not possible to create duplicated ApplicationPermission.
+
+    We show this by trying to create the same ApplicationPermission twice, the first one is correctly created
+    in the database and the response status code is 201. But the second one receives a 409 status in the
+    response and only 1 ApplicationPermission is in the databae and the its id is 1.
+    """
+    user = [UserCreate(is_superuser=True, **user_data)]
+    await insert_objects(user, users_table)
+
+    acl = {"acl": "Allow|role:admin|view"}
+    response = client.post("/application-permissions/", data={"acl": "Allow|role:admin|view"})
+    assert response.status_code == status.HTTP_201_CREATED
+    response = client.post("/application-permissions/", data={"acl": "Allow|role:admin|view"})
+    assert response.status_code == status.HTTP_409_CONFLICT
+
+    count = await database.fetch_all("SELECT COUNT(*) FROM application_permissions")
+    assert count[0][0] == 1
+
+    query = application_permissions_table.select(application_permissions_table.c.id == 1)
+    permission = ApplicationPermission.parse_obj(await database.fetch_one(query))
+
+    assert permission is not None
+    assert permission.acl == acl["acl"]
+    assert permission.id == 1
+
+
+@pytest.mark.asyncio
+@database.transaction(force_rollback=True)
+async def test_create_application_permissions_not_superuser(user_data, client):
+    """
+    Test that is not possible for a non superuser to create permissions.
+
+    This test proves that is not possible to create an ApplicationPermission. We show this by making a
+    request with a non superuser and asserting that the response status code is 401, and that the
+    ApplicationPermission is not created in the database.
+    """
+    user = [UserCreate(is_superuser=False, **user_data)]
+    await insert_objects(user, users_table)
+
+    response = client.post("/application-permissions/", data={"acl": "Allow|role:admin|view"})
+    assert response.status_code == status.HTTP_401_UNAUTHORIZED
+
+    count = await database.fetch_all("SELECT COUNT(*) FROM application_permissions")
+    assert count[0][0] == 0
+
+
+@pytest.mark.asyncio
+@database.transaction(force_rollback=True)
+async def test_create_application_permissions_invalid_acl(user_data, client):
+    """
+    Test that is not possible to create an ApplicationPermission with a bad formatted acl.
+
+    This test proves that is not possible to create the ApplicationPermission with a bad formatted acl. We
+    show this by making a request with a bad acl string and asserting that the response status code is 400
+    and that the ApplicationPermission is not created in the database.
+    """
+    user = [UserCreate(is_superuser=True, **user_data)]
+    await insert_objects(user, users_table)
+
+    response = client.post("/application-permissions/", data={"acl": "Allow|"})
```

Comment:
are there any other bad formats that we should account for? "|view" or "admin|view" or "Allow:|view"

Reply:
Yes, but here we don't need to test them all, since there is already a test for that in the test_schemas for the application_permissions.